stan/math/prim/scal/meta/contains_fvar.hpp

```diff
@@ -3,25 +3,16 @@
 
 #include <stan/math/prim/scal/meta/is_fvar.hpp>
 #include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <stan/math/prim/scal/meta/or.hpp>
 
 namespace stan {
 
 /**
- * Metaprogram to calculate the base scalar return type resulting
- * from promoting all the scalar types of the template parameters.
+ * Metaprogram to determine if any of the
+ * provided types is a fvar.
  */
-template <typename T1, typename T2 = double, typename T3 = double,
-          typename T4 = double, typename T5 = double, typename T6 = double>
-struct contains_fvar {
-  enum {
-    value = is_fvar<typename scalar_type<T1>::type>::value
-            || is_fvar<typename scalar_type<T2>::type>::value
-            || is_fvar<typename scalar_type<T3>::type>::value
-            || is_fvar<typename scalar_type<T4>::type>::value
-            || is_fvar<typename scalar_type<T5>::type>::value
-            || is_fvar<typename scalar_type<T6>::type>::value
-  };
-};
+template <typename... T>
+using contains_fvar = math::or_<is_fvar<typename scalar_type<T>::type>...>;
```

Comment:
I don't understand any of this.  What is the `using` doing here?  And how is that function `is_fvar` getting applied to all the types?  Is there doc for how this works in the language somewhere?  I didn't know what to search for.

Reply:
The `using` is a [type alias declaration](https://en.cppreference.com/w/cpp/language/type_alias). This [stackoverflow anwser](https://stackoverflow.com/a/2795024) describes this way better than I could. This same idea could be used to simply most of the metaprograms ( I would gladly take this task if you guys feel its of any value to do this).

This gets applied to all types by doing what `fold_expressions` does in C++17 ([example](https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/fold-expressions)). This will be greatly simplified once we get there.

Reply:
Thanks for the links.  C++ never fails to amaze me with its odd overloading of syntax.  I'll think of it like a templated typedef.